File: sdk/elixir/lib/rocksky/graph.ex
defmodule Rocksky.Graph do
  @moduledoc "`app.rocksky.graph.*` endpoints."

  alias Rocksky.HTTP

  @doc "Follow an account. Params: `:account` (DID or handle)."
  def follow_account(client, params),
    do: HTTP.procedure(client, "app.rocksky.graph.followAccount", params)

  @doc "Unfollow an account. Params: `:account`."
  def unfollow_account(client, params),
    do: HTTP.procedure(client, "app.rocksky.graph.unfollowAccount", params)

  @doc "Followers of an actor. Params: `:actor`, `:limit`, `:dids`, `:cursor`."
  def get_followers(client, params \\ []),
    do: HTTP.query(client, "app.rocksky.graph.getFollowers", params)

  @doc "Accounts an actor follows. Params: `:actor`, `:limit`, `:dids`, `:cursor`."
  def get_follows(client, params \\ []),
    do: HTTP.query(client, "app.rocksky.graph.getFollows", params)

  @doc "Followers of an actor that the viewer also knows. Params: `:actor`, `:limit`, `:cursor`."
  def get_known_followers(client, params \\ []),
    do: HTTP.query(client, "app.rocksky.graph.getKnownFollowers", params)
end

File: sdk/clojure/src/rocksky/graph.clj
(ns rocksky.graph
  "Endpoints under app.rocksky.graph.*"
  (:require [rocksky.client :as c]))

(defn follow-account
  "Follow an account. Required: `:account` (DID or handle)."
  [client {:keys [account]}]
  (c/procedure client :app.rocksky.graph.followAccount nil {:account account}))

(defn unfollow-account
  "Unfollow an account. Required: `:account`."
  [client {:keys [account]}]
  (c/procedure client :app.rocksky.graph.unfollowAccount nil {:account account}))

(defn get-followers
  "List followers of an actor.

  Required: `:actor`. Optional: `:limit` `:dids` `:cursor`."
  [client {:keys [actor limit dids cursor]}]
  (c/query client :app.rocksky.graph.getFollowers
           {:actor actor :limit limit :dids dids :cursor cursor}))

(defn get-follows
  "List accounts an actor follows.

  Required: `:actor`. Optional: `:limit` `:dids` `:cursor`."
  [client {:keys [actor limit dids cursor]}]
  (c/query client :app.rocksky.graph.getFollows
           {:actor actor :limit limit :dids dids :cursor cursor}))

(defn get-known-followers
  "List followers of `:actor` that the viewer also follows.

  Required: `:actor`. Optional: `:limit` `:cursor`."
  [client {:keys [actor limit cursor]}]
  (c/query client :app.rocksky.graph.getKnownFollowers
           {:actor actor :limit limit :cursor cursor}))
